<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  

[CEMBA Data Processing](#cemba-data-processing)
  - [Giant workflow](#giant-workflow)
  - [Processing Pipeline](#processing-pipeline)
    - [1. Input of the pipeline](#1-input-of-the-pipeline)
      - [1.1 Methylation Data](#11-methylation-data)
      - [1.2 Cell Meta-Data](#12-cell-meta-data)
      - [1.3 Genome and Gene Definition](#13-genome-and-gene-definition)
    - [2. Group by region](#2-group-by-region)
      - [2.1 Methylation Types](#21-methylation-types)
        - [mCG](#mcg)
        - [mCH (also known as non-CG methylation)](#mch-also-known-as-non-cg-methylation)
      - [2.2 Region Types](#22-region-types)
      - [ChromBin](#chrombin)
      - [Gene](#gene)
      - [[TODO] DMR](#todo-dmr)
    - [3. mC matrix preprocess](#3-mc-matrix-preprocess)
    - [4. Cell Level Computation](#4-cell-level-computation)
    - [5. Marker Gene](#5-marker-gene)
  - [Project Structure](#project-structure)
  - [CEMBA Data Format](#cemba-data-format)
    - [Cell Meta-data format](#cell-meta-data-format)
    - [ALLC format](#allc-format)
    - [Region ALLC format](#region-allc-format)
    - [mC matrix (not shared on dropbox)](#mc-matrix-not-shared-on-dropbox)
      - [cov base call matrix (for group of cell, certain mC type)](#cov-base-call-matrix-for-group-of-cell-certain-mc-type)
      - [mc base call matrix (for group of cell, certain mC type)](#mc-base-call-matrix-for-group-of-cell-certain-mc-type)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# CEMBA Data Processing
**Data Preprocessing, Cell Clustering, Marker Gene Identification**

## Giant workflow
<img src="./CEMBA_workflow.svg">

## Processing Pipeline
### 1. Input of the pipeline
#### Methylation Data
All the analysis start from ALLC file, which is generated by [methylpy](https://github.com/yupenghe/methylpy).
Each ALLC file contains the base level methylation information for each cell. See bellow for ALLC file format.
#### Cell Meta-Data
See bellow for format.
#### Genome and Gene Definition
Genome: mouse mm10
Gene and other features: [GENCODE vM16](https://www.gencodegenes.org/mouse_releases/16.html)


### 2. Group by region
#### Methylation Types
##### mCG
where base context in ALLC is CGN (N for A|T|C|G).
##### mCH (also known as non-CG methylation)
where base context in ALLC is CHN (H for A|T|C).
see [here](https://www.ncbi.nlm.nih.gov/pubmed/26077819) for the biological difference of these two kinds of methylation types.

#### Region Types
For Methylation data, base level information is too sparse to analysis. So we should reduce the information into region level first. There are 3 major region types we can use.

1. Chromosome bins with fixed length (ChromBin)
2. Gene body methylation (Gene)
3. Computed regions like differential methylated region (DMR)

Currently, we only prepared the ChromBin and Gene regions. 

#### ChromBin
Fixed chromosomal 100kb bins from mm10
See the BED file here: [Put BED file link here]
This region level information will be used in **tSNE and clustering**. Because it contains all the information across genome. Different size of the ChromBin can give out different result, currently, we only provide 100kb bins, which is suitable for most analysis.

#### Gene
Full Gene body generated from [GENCODE vM16](https://www.gencodegenes.org/mouse_releases/16.html)
See the BED file here: [Put BED file link here]
This region level information (mCH) will be used in finding marker gene.

#### [TODO] DMR
[TODO]

### 3. mC matrix preprocess 
In all the matrix starting from this step, the row is cell, the column is feature, where feature can be region mCH/mCG%, PCA/tSNE components, Clusters etc.

#### Base Call Filtering
**base call cutoff**:

- For mCH, cutoff is 100;
- For mCG, cutoff is 10, because in genome, mCG base number is much smaller than mCH, the cutoff is reduced correspondingly.

In regions coverage matrix, for each value < cutoff, treat it as NA.

#### Cell Coverage Filtering
**cell coverage cutoff**: 50%
Remove cells having >50% NAs after base call cutoff

#### Region Coverage Filtering
**region coverage cutoff**
Remove cells having > cutoff% NAs after base call cutoff

- For ChromBin mCH and mCG region: 95%, because these regions will be imputed subsequently, only <5% values are imputed after this filter.
- For Gene Body mCH and mCG region: 20%, because we will not impute Gene Body mC%, just using this to visualize or find markers.

#### Cell Overall mC Normalizing
The average mC% can be very different between cells, so we calculated cell overall mCH/mCG% across whole genome and normalize each cell's (row's) value by dividing the overall mC%.

#### Imputation
For each region (column), NA values are imputed by the mean of region's (columns's) mCH/mCG% respectively. At most 5% of the values are imputed.

#### Feature Normalizing
The mCH% and mCG% have very different scale, so in order to properly combine them, we normalized (using their IQR) and centered each region (column), see reference [here](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.RobustScaler.html).

#### Output of This Part
After mC matrix preprocess, we got a mC% matrix, where row is cell, column is feature. Both column and row are filtered by coverage, and then normalized and imputed. This matrix is used as input for next step.

### 4. Cell Level Computation
#### PCA
We use PCA to decompose the input matrix into 50 components, see reference [here](http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html).
Then we filtered out outliers based on each data point's local density with respect to its neighbors, see reference [here](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.LocalOutlierFactor.html#sklearn.neighbors.LocalOutlierFactor). 
After removing outliers, we redo the PCA and get the PCA matrix.

#### tSNE
We do tSNE, with parameters n_components=2, perplexity=50, other parameters set to default, see reference [here](http://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html)

#### Cluster
##### Shared KNN Graph
We first calculated KNN for each cell, where K=20. Then the edge weight of the shared KNN graph is defined as the Jaccard coefficient of the two nodes' KNN set, see reference [here](https://github.com/jacoblevine/PhenoGraph).

##### Louvian Method
After building the graph, we identified communities (clusters) by using the Louvian method, see reference [here](https://github.com/taynaud/python-louvain).

We first identified 66 clusters using Resolution=0.2, then we merge these clusters under different resolutions (0.3, 0.5, 0.7, 0.9, 1.0) to get a hierarchal cluster tree.

##### Known Brain Region
Known brain regions are labeled based on canonical markers.

For excitatory neuron: 

- L2/3: Cux2+; Rorb-
- L4: Cux2+; Rorb+
- L5a: Cux2-; Rorb+
- L5b: Bcl6+
- L6: Tle4+
- DL: other Gad1- cells

For inhibitory neuron:

- Vip: Vip+
- Ndnf: Ndnf+
- Pv: Pvalb+
- Sst: Sst+

For Glia:

- Low overall mCH

#### Output of This Part
PCA matrix: row is cell, column is PCA components
tSNE matrix: row is cell, column is tSNE components
Cluster matrix: row is cell, column is cluster solutions under different resolution or known region labeling.

### 5. Marker Gene
Marker gene is identified based on pairwise quantile compare of the cluster's gene mCH%. 
[TODO, add details]


## Python Package Structure
- `cemba_data.tools` scripts for data cleaning and computation.
- `cemba_data.db` scripts for data storage into MongoDB
- `cemba_data.working` preparing helpers for running the scripts.

## CEMBA Data Format

### Cell Meta-data format
[TODO]

### ALLC format 
(for each cell, not shared on dropbox)

| chrom | pos | strand | context | mc | cov | p |
| --- | --- | --- | --- | --- | --- | --- |
| 1 | 24147 | - | CTA | 0 | 1 | 1 |
| 2 | 24148 | - | CCT | 0 | 1 | 1 |
| 3 | 24150 | - | CTC | 0 | 1 | 1 |
| X | 24155 | - | CAT | 0 | 1 | 1 |
| Y | 24156 | - | CCA | 0 | 1 | 1 |
| M | 24159 | - | CTA | 0 | 1 | 1 |
- chrom: no chr; may include L (lambda DNA)
- pos: mm10 position start from 1
- strand
- context: CNN format, showing 2 base after the C
- mc: count of methylated base
- cov: count of total base at this position
- p: not used in single cell data

### Region ALLC format 
(for each cell, not shared on dropbox)

| id | data_type | CAA | CAT | CAC | CAG | CTA | CTT | CTC | CTG | CCA | CCT | CCC | CCG | CGA | CGT | CGC | CGG |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| - | mc | 1 | 0 | 1 | 1 | 1 | 0 | 0 | 0 | 0 | 1 | 0 | 0 | 0 | 0 | 0 | 0 |
| - | cov | 122 | 147 | 77 | 116 | 124 | 109 | 68 | 126 | 110 | 102 | 67 | 56 | 56 | 84 | 63 | 66 |
| - | mc | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 0 | 1 | 2 |
| - | cov | 9 | 7 | 6 | 11 | 3 | 7 | 5 | 11 | 10 | 7 | 3 | 1 | 1 | 1 | 1 | 2 |
| - | mc | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
| - | cov | 7 | 5 | 2 | 2 | 3 | 8 | 2 | 1 | 3 | 3 | 1 | 0 | 0 | 0 | 0 | 0 |
| - | mc | 0 | 2 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 2 | 0 | 0 | 2 | 13 | 6 | 2 |
| - | cov | 57 | 61 | 73 | 77 | 51 | 80 | 81 | 66 | 99 | 74 | 74 | 5 | 2 | 15 | 6 | 2 |
| - | mc | 2 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 4 | 5 | 1 | 1 |
| - | cov | 10 | 11 | 9 | 17 | 7 | 16 | 7 | 10 | 17 | 10 | 11 | 6 | 4 | 5 | 1 | 2 |
- id: for gene, ENSEMBLE id; for chrom bin: chrom + bin_number
- shape: number of columns and rows are not fixed
- cell: raw base call, only filtered by max_cov_cutoff (default=2, means for a given pos in ALLC, cod <= 2)

### mC matrix 
(not shared on dropbox)
#### cov base call matrix 
(for group of cell, certain mC type)

|  | 3_0 | 3_1 | 3_2 | ... | 16_976 | 16_977 | 16_978 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 0 | 0 | 0 | 0 | ... | 1532 | 1230 | 1087 |
| 1 | 0 | 0 | 0 | ... | 1879 | 1929 | 2680 |
| … | … | … | … | … | … | … | … |
| 98 | 0 | 0 | 0 | ... | 1765 | 963 | 744 |
| 99 | 0 | 0 | 0 | ... | 865 | 2115 | 1588 |
| … | … | … | … | … | … | … | … |

#### mc base call matrix 
(for group of cell, certain mC type)

|  | 3_0 | 3_1 | 3_2 | ... | 16_976 | 16_977 | 16_978 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 0 | 0 | 0 | 0 | ... | 54 | 40 | 58 |
| 1 | 0 | 0 | 0 | ... | 70 | 70 | 102 |
| … | … | … | … | … | … | … | … |
| 98 | 0 | 0 | 0 | ... | 37 | 14 | 18 |
| 99 | 0 | 0 | 0 | ... | 30 | 89 | 54 |
| … | … | … | … | … | … | … | … |



