{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare Input Files for Pipeline\n",
    "\n",
    "## FASTQ Requirements\n",
    "- FASTQ file should be generated by SampleSheet made from previous step. This is to make sure the FASTQ names can be automatically parsed to get the uid, lane, read_type information.\n",
    "- If the SampleSheet is not generated by previous step, name may not support automatic parse. You need to provide the FASTQ dataframe by yourself. See the documentation of making FASTQ dataframe bellow.\n",
    "- The supported default name pattern of FASTQ file should be \n",
    "```python\n",
    "*_, plate1, plate2, multi_field = path.name.split('-')\n",
    "plate_pos, _, lane, read_type, _ = multi_field.split('_')\n",
    "```\n",
    "\n",
    "## Input Files\n",
    "In order to run the pipeline for a single cell library, you need to have 2 things ready:\n",
    "1. FASTQ files as described above for the whole library\n",
    "2. mapping_config.ini for mapping parameters\n",
    "3. (optional) fastq_dataframe.csv for fastq information. This file can be automatically generated by yap, if the fastq is demultiplexed by SampleSheet made from `yap make-sample-sheet`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mapping Modes\n",
    "yap support several different mapping mode for different experiments, which is controlled by mapping_config.ini as described bellow.\n",
    "\n",
    "\n",
    "#### snmC-seq2 (mc)\n",
    "- Normal snmC-seq2 library\n",
    "\n",
    "#### snmCT-seq (mct)\n",
    "- snmCT-seq library, each cell contain mixed reads from DNA and RNA\n",
    "- The major differences are:\n",
    "    - Need to do STAR mapping\n",
    "    - Filter bismark BAM file to get DNA reads\n",
    "    - Filter STAR BAM file to get RNA reads\n",
    "\n",
    "#### snmC-seq + NOMe treatment (nome)\n",
    "- snmC-seq with NOMe treatment, where the GCH contain open chromatin information, HCN contain normal methylation information.\n",
    "- The major differences are:\n",
    "    - Add additional one base in the context column of ALLC\n",
    "\n",
    "####  snmCT-seq + NOMe treatment (mct-nome)\n",
    "- snmCT-seq with NOMe treatment, each cell contain mixed reads from DNA and RNA, and in DNA reads, the GCH contain open chromatin information, HCN contain normal methylation.\n",
    "- The major differences are:\n",
    "    - Need to do STAR mapping\n",
    "    - Filter bismark BAM file to get DNA reads\n",
    "    - Filter STAR BAM file to get RNA reads\n",
    "    - Add additional one base in the context column of ALLC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mapping_config.ini\n",
    "\n",
    "### What is mapping_config.ini file?\n",
    "\n",
    "- It's a place gather all adjustable parameters of mapping pipeline into a single file in [INI format](https://en.wikipedia.org/wiki/INI_file), so you don't need to put in 100 parameters in a shell command...\n",
    "\n",
    "- INI format is super simple:\n",
    "    ```\n",
    "    ; comment start with semicolon\n",
    "    [section1]\n",
    "    key1=value1\n",
    "    key2=value2\n",
    "    \n",
    "    [section2]\n",
    "    key1=value1\n",
    "    key2=value2\n",
    "    ```\n",
    "- Currently, the pipeline don't allow to change the sections and keys, so just change values according to your needs.\n",
    "\n",
    "### How to prepare mapping_config.ini file?\n",
    "You can print out the default config file, save it to your own place and modify the value.\n",
    "```shell\n",
    "# MODE should be in mc, nome, mct, mct-nome, depending on the library type. Default is mc\n",
    "# see different mapping mode bellow\n",
    "yap default-mapping-config --mode MODE\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ dataframe (optional)\n",
    "**NOTE: This file can be automatically generated by yap, if the fastq is demultiplexed by SampleSheet made from `yap make-sample-sheet`**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### What is FASTQ dataframe?\n",
    "- A tab separated table contain all information of a library of FASTQ files.\n",
    "- First line is header, each other line is information of a FASTQ file.\n",
    "- IMPORTANT: must have 4 columns, (order of column doesn't matter, but the column name must be the same as following):\n",
    "    - lane: sequencing lane where the fastq come from, e.g. L001, L002, L003, L004...\n",
    "    - uid: An unique id that corresponding to each illumina i5 i7 index.\n",
    "    - read_type: read1 or read2, e.g. R1, R2\n",
    "    - fastq_path: Absolute path of a FASTQ file.\n",
    "\n",
    "### How to prepare FASTQ dataframe?\n",
    "- Because different project may have different naming pattern, you should figure out which part of the file name is lane, uid and read_type, then make these information for each FASTQ file in a table like bellow.\n",
    "- This may seems trivial, but once you done these, you can save the code for all subsequent experiments following same name pattern.\n",
    "- so it is HIGHLY RECOMMENDED THAT YOU DO NOT CHANGE FASTQ NAME PATTERN THROUGHOUT YOUR PROJECT."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTQ dataframe example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-14T05:17:02.934836Z",
     "start_time": "2019-09-14T05:17:02.908216Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This FASTQ dataframe have 3072 rows (not include the first header row) for 3072 FASTQ file, which is from a standard CEMBA snmC-seq experiment using 8 384-well plates.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lane</th>\n",
       "      <th>uid</th>\n",
       "      <th>read_type</th>\n",
       "      <th>fastq_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L001</td>\n",
       "      <td>CEMBA181022-6B-1-CEMBA181022-6B-2-A1</td>\n",
       "      <td>R1</td>\n",
       "      <td>/gale/oberon/data11/181119_A00280_0033_BHCHYKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L001</td>\n",
       "      <td>CEMBA181022-6B-1-CEMBA181022-6B-2-A1</td>\n",
       "      <td>R2</td>\n",
       "      <td>/gale/oberon/data11/181119_A00280_0033_BHCHYKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L001</td>\n",
       "      <td>CEMBA181022-6B-1-CEMBA181022-6B-2-A2</td>\n",
       "      <td>R1</td>\n",
       "      <td>/gale/oberon/data11/181119_A00280_0033_BHCHYKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L001</td>\n",
       "      <td>CEMBA181022-6B-1-CEMBA181022-6B-2-A2</td>\n",
       "      <td>R2</td>\n",
       "      <td>/gale/oberon/data11/181119_A00280_0033_BHCHYKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L001</td>\n",
       "      <td>CEMBA181022-6B-1-CEMBA181022-6B-2-A3</td>\n",
       "      <td>R1</td>\n",
       "      <td>/gale/oberon/data11/181119_A00280_0033_BHCHYKD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   lane                                   uid read_type  \\\n",
       "0  L001  CEMBA181022-6B-1-CEMBA181022-6B-2-A1        R1   \n",
       "1  L001  CEMBA181022-6B-1-CEMBA181022-6B-2-A1        R2   \n",
       "2  L001  CEMBA181022-6B-1-CEMBA181022-6B-2-A2        R1   \n",
       "3  L001  CEMBA181022-6B-1-CEMBA181022-6B-2-A2        R2   \n",
       "4  L001  CEMBA181022-6B-1-CEMBA181022-6B-2-A3        R1   \n",
       "\n",
       "                                          fastq_path  \n",
       "0  /gale/oberon/data11/181119_A00280_0033_BHCHYKD...  \n",
       "1  /gale/oberon/data11/181119_A00280_0033_BHCHYKD...  \n",
       "2  /gale/oberon/data11/181119_A00280_0033_BHCHYKD...  \n",
       "3  /gale/oberon/data11/181119_A00280_0033_BHCHYKD...  \n",
       "4  /gale/oberon/data11/181119_A00280_0033_BHCHYKD...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_table('fastq_dataframe.tsv.gz', index_col=None)[['lane', 'uid', 'read_type', 'fastq_path']]\n",
    "print(f'This FASTQ dataframe have {df.shape[0]} rows (not include the first header row) for {df.shape[0]} FASTQ file, which is from a standard CEMBA snmC-seq experiment using 8 384-well plates.')\n",
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "source": [],
    "metadata": {
     "collapsed": false
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}